<div id="overview">
  <h2><a name="overview" href="#overview">Brief Overview and Quality Check</a></h2>

  <h3><a name="number_of_reads" href="#number_of_reads">Number of reads</a></h3>
  <table>
    <thead>
      <tr>
        <th>Sample Name</th>
        <th>Total Reads</th>
        <th>Remaining Reads After Processing</th>
        <th>Uniquely Mappable Reads To Genome</th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <th>CLIP-35L33G</th>
        <td>25,212,399</td>
        <td>18,552,847 (83.2%)</td>
        <td>13,153,901 (52.9%)</td>
      </tr>
      <tr>
        <th>CLIP-2J3</th>
        <td>25,212,399</td>
        <td>18,552,847 (83.2%)</td>
        <td>13,153,901 (52.9%)</td>
      </tr>
      <tr>
        <th>RNAseq-uninfected-1</th>
        <td>25,212,399</td>
        <td>18,552,847 (83.2%)</td>
        <td>13,153,901 (52.9%)</td>
      </tr>
    </tbody>
  </table>

  <h3><a name="assignment_profile" href="#assignment_profile">Read Origin Assignment Profiles</a></h3>
  <p>
    <img src="report/plots/read-class-proportions.png"/>
    <a class="link_pdf" href="report/plots/read-class-proportions.pdf">Vector Image (pdf)</a>
    <a class="link_png" href="report/plots/read-class-proportions.png">Bitmap Image (png)</a>
    <a class="link_csv" href="report/tables/read-class-proportions.reads.csv">Read Count Table (csv)</a>
    <a class="link_csv" href="report/tables/read-class-proportions.percentages.csv">Percentage Table (csv)</a>
  </p>

  <table>
    <thead>
      <tr>
        <th>Sample Name</th>
        <th>rRNA</th>
        <th>CDS</th>
        <th>5'UTR</th>
        <th>3'UTR</th>
        <th>intron</th>
        <th>miRNA</th>
        <th>tRNA</th>
        <th>Repeats</th>
        <th>Others</th>
        <th>Unannotated</th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <th>CLIP-35L33G</th>
        <td>25,212,399 (12.5%)</td>
        <td>18,552,847 (83.2%)</td>
        <td>13,153,901 (52.9%)</td>
        <td>25,212,399 (12.5%)</td>
        <td>18,552,847 (83.2%)</td>
        <td>13,153,901 (52.9%)</td>
        <td>25,212,399 (12.5%)</td>
        <td>18,552,847 (83.2%)</td>
        <td>13,153,901 (52.9%)</td>
        <td>13,153,901 (52.9%)</td>
      </tr>
      <tr>
        <th>CLIP-2J3</th>
        <td>25,212,399 (12.5%)</td>
        <td>18,552,847 (83.2%)</td>
        <td>13,153,901 (52.9%)</td>
        <td>25,212,399 (12.5%)</td>
        <td>18,552,847 (83.2%)</td>
        <td>13,153,901 (52.9%)</td>
        <td>25,212,399 (12.5%)</td>
        <td>18,552,847 (83.2%)</td>
        <td>13,153,901 (52.9%)</td>
        <td>13,153,901 (52.9%)</td>
      </tr>
      <tr>
        <th>CLIP-46020</th>
        <td>25,212,399 (12.5%)</td>
        <td>18,552,847 (83.2%)</td>
        <td>13,153,901 (52.9%)</td>
        <td>25,212,399 (12.5%)</td>
        <td>18,552,847 (83.2%)</td>
        <td>13,153,901 (52.9%)</td>
        <td>25,212,399 (12.5%)</td>
        <td>18,552,847 (83.2%)</td>
        <td>13,153,901 (52.9%)</td>
        <td>13,153,901 (52.9%)</td>
      </tr>
      <tr>
        <th>RNAseq-uninfected-1</th>
        <td>25,212,399 (12.5%)</td>
        <td>18,552,847 (83.2%)</td>
        <td>13,153,901 (52.9%)</td>
        <td>25,212,399 (12.5%)</td>
        <td>18,552,847 (83.2%)</td>
        <td>13,153,901 (52.9%)</td>
        <td>25,212,399 (12.5%)</td>
        <td>18,552,847 (83.2%)</td>
        <td>13,153,901 (52.9%)</td>
        <td>13,153,901 (52.9%)</td>
      </tr>
    </tbody>
  </table>
</div>
{#-
vim: sts=2 sw=2 et ft=htmljinja
-#}
