<h2><a name="binding_characteristics" href="#binding_characteristics">Binding Characteristics</a></h2>
<div id="binding_characteristics">
  <h3><a name="motif_logos" href="#motif_logos">Sequence Motif Logos</a></h3>
  <p>
    These WebLogos represent enriched patterns of flanking sequences around confident
    binding sites. The motifs are aligned just to center the identified binding sites
    at the single nucleotide level, so this may underestimate loose patterns with
    variable distances between recognized bases.
  </p>
  <table>
    <thead>
      <tr>
        <th>Sample</th>
        <th>Genomic Sequences</th>
        <th>Spliced Sequences</th>
      </tr>
    </thead>
    <tbody>
      {% for sample in samples('CLIP') %}
      <tr>
        <th>{{ sample.label }}</th>
        <td>
          <p class="buttons">
            <a href="report/plots/binding-site/logo-{{ sample.label }}-genome-entropy-probability.pdf" class="link_pdf">PDF</a>
            <a href="report/plots/binding-site/logo-{{ sample.label }}-genome-entropy-probability.png" class="link_png">PNG</a>
            <a href="#" class="link_csv">Show more</a>
          </p>
          <img src="report/plots/binding-site/logo-{{ sample.label }}-genome-entropy-probability.png"/>
        </td>
        <td>
          <p class="buttons">
            <a href="report/plots/binding-site/logo-{{ sample.label }}-transcriptome-entropy-probability.pdf" class="link_pdf">PDF</a>
            <a href="report/plots/binding-site/logo-{{ sample.label }}-transcriptome-entropy-probability.png" class="link_png">PNG</a>
            <a href="#" class="link_csv">Show more</a>
          </p>
          <img src="report/plots/binding-site/logo-{{ sample.label }}-transcriptome-entropy-probability.png"/>
        </td>
      </tr>
      {% endfor %}
    </tbody>
  </table>

  <h3><a name="secondary_structure" href="#secondary_structure">Secondary Structure Preference</a></h3>
  <table>
    <thead>
      <tr>
        <th>Sample</th>
        <th>Genomic Sequences</th>
        <th>Spliced Sequences</th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <th>CLIP-35L33G</th>
        <td>PLOTS HERE (many)</td>
        <td>PLOTS HERE (many)</td>
      </tr>
      <tr>
        <th>CLIP-2J3</th>
        <td>PLOTS HERE (many)</td>
        <td>PLOTS HERE (many)</td>
      </tr>
      <tr>
        <th>RNAseq-uninfected-1</th>
        <td>PLOTS HERE (many)</td>
        <td>PLOTS HERE (many)</td>
      </tr>
    </tbody>
  </table>

  <h3><a name="enriched_nmer" href="#enriched_nmer">Enriched N-mer Sequences</a></h3>
  <table>
    <thead>
      <tr>
        <th>Sample</th>
        <th>Genomic Sequences</th>
        <th>Spliced Sequences</th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <th>CLIP-35L33G</th>
        <td>PLOTS HERE (many)</td>
        <td>PLOTS HERE (many)</td>
      </tr>
      <tr>
        <th>CLIP-2J3</th>
        <td>PLOTS HERE (many)</td>
        <td>PLOTS HERE (many)</td>
      </tr>
      <tr>
        <th>RNAseq-uninfected-1</th>
        <td>PLOTS HERE (many)</td>
        <td>PLOTS HERE (many)</td>
      </tr>
    </tbody>
  </table>

  <h3><a name="binding_alternatives" href="#binding_alternatives">
    Alternative Binding Pattern Discovery</a></h3>
  <p>
    These are pattern analysis results using external motif discovery tools for
    flanking sequences around confident binding sites. Click a link to view
    a result.
  </p>

  <table>
    <thead>
      <tr>
        <th>Sample</th>
        <th>MEME-ChIP</th>
        <th>GLAM2</th>
        <th>PhyloGibbs</th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <th>CLIP-35L33G</th>
        <td>
          <a href="report/meme/CLIP-35L33G">Genomic</a><br/>
          <a href="report/meme/CLIP-35L33G">Spliced</a>
        </td>
        <td>
          <a href="report/meme/CLIP-35L33G">Genomic</a><br/>
          <a href="report/meme/CLIP-35L33G">Spliced</a>
        </td>
        <td>
          <a href="report/meme/CLIP-35L33G">Genomic</a><br/>
          <a href="report/meme/CLIP-35L33G">Spliced</a>
        </td>
      </tr>
      <tr>
        <th>CLIP-2J3</th>
        <td>
          <a href="report/meme/CLIP-35L33G">Genomic</a><br/>
          <a href="report/meme/CLIP-35L33G">Spliced</a>
        </td>
        <td>
          <a href="report/meme/CLIP-35L33G">Genomic</a><br/>
          <a href="report/meme/CLIP-35L33G">Spliced</a>
        </td>
        <td>
          <a href="report/meme/CLIP-35L33G">Genomic</a><br/>
          <a href="report/meme/CLIP-35L33G">Spliced</a>
        </td>
      </tr>
      <tr>
        <th>RNAseq-uninfected-1</th>
        <td>
          <a href="report/meme/CLIP-35L33G">Genomic</a><br/>
          <a href="report/meme/CLIP-35L33G">Spliced</a>
        </td>
        <td>
          <a href="report/meme/CLIP-35L33G">Genomic</a><br/>
          <a href="report/meme/CLIP-35L33G">Spliced</a>
        </td>
        <td>
          <a href="report/meme/CLIP-35L33G">Genomic</a><br/>
          <a href="report/meme/CLIP-35L33G">Spliced</a>
        </td>
      </tr>
    </tbody>
  </table>

</div>
{#-
vim: sts=2 sw=2 et ft=htmljinja
-#}
