<h2><a name="data_files" href="#data_files">Data Files for Subsequent Analyses</a></h2>
<div id="data_files">
  <h3><a name="data_sequence_files" href="#data_sequence_files">Processed Sequence Files</a></h3>
  <table>
    <thead>
      <tr>
        <th>Sample</th>
        <th>File</th>
        <th>Size</th>
        <th>Format</th>
      </tr>
    </thead>
    <tbody>
      {% for sample in samples(('CLIP', 'RNAseq')) %}
      <tr>
        <th>{{ sample.label }}</th>
        <td>
          <a href="sequences/{{ sample.label }}.filtered.fa">{{ sample.label }}.fa</a><br/>
          <a href="sequences/{{ sample.label }}.filtered.fq.gz">{{ sample.label }}.fq.gz</a>
        </td>
        <td>
          {{ filesize('sequences/%s.filtered.fa'|format(sample.label))|human_readable_bytes }}<br/>
          {{ filesize('sequences/%s.filtered.fq.gz'|format(sample.label))|human_readable_bytes }}
        </td>
        <td><a href="http://genetics.bwh.harvard.edu/pph/FASTA.html">Fasta</a> (collapsed)<br/><a href="http://maq.sourceforge.net/fastq.shtml">FastQ</a></td>
      </tr>
      {% endfor %}
    </tbody>
  </table>

  <h3><a name="data_alignment_files" href="#data_alignment_files">Alignment Files</a></h3>
  <table>
    <thead>
      <tr>
        <th>Sample</th>
        <th>File</th>
        <th>Size</th>
        <th>Format</th>
      </tr>
    </thead>
    <tbody>
      {% for sample in samples(('CLIP', 'RNAseq')) %}
      <tr>
        <th>{{ sample.label }}</th>
        <td>
          <a href="alignments/{{ sample.label }}.sorted.bam">{{ sample.label }}.sorted.bam</a><br/>
          <a href="alignments/{{ sample.label }}.sorted.bam.bai">{{ sample.label }}.sorted.bam.bai</a>
        </td>
        <td>
          {{ filesize('alignments/%s.sorted.bam'|format(sample.label))|human_readable_bytes }}<br/>
          {{ filesize('alignments/%s.sorted.bam.bai'|format(sample.label))|human_readable_bytes }}
        </td>
        <td><a href="https://genome.ucsc.edu/FAQ/FAQformat.html#format5.1">bam</a> (sorted)<br/>bam index</td>
      </tr>
      {% endfor %}
    </tbody>
  </table>

  <h3><a name="data_annotations" href="#data_annotations">Sequence Annotation Files</a></h3>
  <table>
    <thead>
      <tr>
        <th>Sample</th>
        <th>File</th>
        <th>Size</th>
        <th>Format</th>
      </tr>
    </thead>
    <tbody>
      {% for sample in samples(('CLIP', 'RNAseq')) %}
      <tr>
        <th>{{ sample.label }}</th>
        <td>
          <a href="annotations/{{ sample.label }}.anno.gz">{{ sample.label }}.anno.gz</a>
        </td>
        <td>{{ filesize('annotations/%s.anno.gz'|format(sample.label))|human_readable_bytes }}</td>
        <td><a href="report/help/annotations.html" target="_new">(help)</a></td>
      </tr>
      {% endfor %}
    </tbody>
  </table>

  <h3><a name="data_site_catalogs" href="#data_site_catalogs">
    Nucleotide-resolution Statistics for Binding Site Identification</a></h3>
  <table>
    <thead>
      <tr>
        <th>Sample</th>
        <th>File</th>
        <th>Size</th>
        <th>Format</th>
      </tr>
    </thead>
    <tbody>
      {% for sample in samples('CLIP') %}
      <tr>
        <th>{{ sample.label }}</th>
        <td>
          <a href="erroranalyses/{{ sample.label }}.catalog.gz">{{ sample.label }}.catalog.gz</a>
        </td>
        <td>{{ filesize('erroranalyses/%s.catalog.gz'|format(sample.label))|human_readable_bytes }}</td>
        <td><a href="report/help/site_catalogs.html" target="_new">(help)</a></td>
      </tr>
      {% endfor %}
    </tbody>
  </table>

  <h3><a name="data_site_fdr_estimation" href="#data_site_fdr_estimation">
    Binding Site Score Cutoff Tables for Desired FDR Levels</a></h3>
  <table>
    <thead>
      <tr>
        <th>Sample</th>
        <th>File</th>
        <th>Size</th>
        <th>Format</th>
      </tr>
    </thead>
    <tbody>
      {% for sample in samples('CLIP') %}
      <tr>
        <th>{{ sample.label }}</th>
        <td>
          {% for method in 'del mod moddel entropy t2c'.split() %}
          <a href="erroranalyses/{{ sample.label }}.{{ method }}.fdr_cutoffs">{{ sample.label }}.{{ method }}.fdr_cutoffs</a><br/>
          {% endfor %}
        </td>
        <td>
          {% for method in 'del mod moddel entropy t2c'.split() %}
          {{ filesize('erroranalyses/%s.%s.fdr_cutoffs'|format(sample.label, method))|human_readable_bytes }}
          {%- if not loop.last %}<br/>{% endif -%}
          {% endfor %}
        </td>
        <td><a href="report/help/site_catalogs.html" target="_new">(help)</a></td>
      </tr>
      {% endfor %}
    </tbody>
  </table>

  <h3><a name="data_piranha_peaks" href="#data_piranha_peaks">
    Alternative Binding Sites Calls (Piranha)</a></h3>
  <table>
    <thead>
      <tr>
        <th>Sample</th>
        <th>File</th>
        <th>Size</th>
        <th>Format</th>
      </tr>
    </thead>
    <tbody>
      {% for sample in samples('CLIP') %}
      <tr>
        <th>{{ sample.label }}</th>
        <td>
          <a href="peaks/{{ sample.label }}.piranha.gz">{{ sample.label }}.piranha.gz</a>
        </td>
        <td>{{ filesize('peaks/%s.piranha.gz'|format(sample.label))|human_readable_bytes }}</td>
        <td><a href="https://genome.ucsc.edu/FAQ/FAQformat.html#format1">gzipped bed</a></td>
      </tr>
      {% endfor %}
    </tbody>
  </table>

  <h3><a name="data_enriched_transcripts" href="#data_enriched_transcripts">
    Target Transcript List Based on Tag Abundance</a></h3>
  <table>
    <thead>
      <tr>
        <th>Pair Set</th>
        <th>File</th>
        <th>Size</th>
        <th>Format</th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <th>CLIP-35L33G</th>
        <td>
          <a href="stats/CLIP-35L33G.enrichment.csv">CLIP-35L33G.enrichment.csv</a>
        </td>
        <td>59.1 MiB</td>
        <td><a href="https://genome.ucsc.edu/FAQ/FAQformat.html#format1">csv</a></td>
      </tr>
      <tr>
        <th>CLIP-2J3</th>
        <td>
          <a href="stats/CLIP-2J3.enrichment.csv">CLIP-2J3.enrichment.csv</a>
        </td>
        <td>9.1 MiB</td>
        <td><a href="https://genome.ucsc.edu/FAQ/FAQformat.html#format1">csv</a></td>
      </tr>
    </tbody>
  </table>

  <h3><a name="data_genome_arrays" href="#data_genome_arrays">
    Genome-Spanning Arrays of CLIP Tag Counts (and more)</a></h3>
  <table>
    <thead>
      <tr>
        <th>Sample</th>
        <th>File</th>
        <th>Size</th>
        <th>Format</th>
      </tr>
    </thead>
    <tbody>
      {% for sample in samples('CLIP') %}
      <tr>
        <th>{{ sample.label }}</th>
        <td>
          <a href="quantitations/{{ sample.label }}.gspace">{{ sample.label }}.gspace</a> (dir)
        </td>
        <td>{{ filesize('quantitations/%s.gspace'|format(sample.label))|human_readable_bytes }}</td>
        <td><a href="https://genome.ucsc.edu/FAQ/FAQformat.html#format1">(help)</a></td>
      </tr>
      {% endfor %}
    </tbody>
  </table>
</div>
{#-
vim: sts=2 sw=2 et ft=htmljinja
-#}
