<h2><a name="data_files" href="#data_files">Data Files for Subsequent Analyses</a></h2>
<div id="data_files">
  <h3><a name="data_sequence_files" href="#data_sequence_files">Processed Sequence Files</a></h3>
  <table>
    <thead>
      <tr>
        <th>Sample</th>
        <th>File</th>
        <th>Size</th>
        <th>Format</th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <th>CLIP-35L33G</th>
        <td>
          <a href="sequences/CLIP-35L33G.filtered.fa.gz">CLIP-35L33G.fa.gz</a><br/>
          <a href="sequences/CLIP-35L33G.filtered.fq.gz">CLIP-35L33G.fq.gz</a>
        </td>
        <td>7.4 MiB<br/>3.7 MiB</td>
        <td><a href="http://genetics.bwh.harvard.edu/pph/FASTA.html">Fasta</a> (collapsed)<br/><a href="http://maq.sourceforge.net/fastq.shtml">FastQ</a></td>
      </tr>
      <tr>
        <th>CLIP-2J3</th>
        <td>
          <a href="sequences/CLIP-2J3.filtered.fa.gz">CLIP-2J3.fa.gz</a><br/>
          <a href="sequences/CLIP-2J3.filtered.fq.gz">CLIP-2J3.fq.gz</a>
        </td>
        <td>7.4 MiB<br/>3.7 MiB</td>
        <td><a href="http://genetics.bwh.harvard.edu/pph/FASTA.html">Fasta</a> (collapsed)<br/><a href="http://maq.sourceforge.net/fastq.shtml">FastQ</a></td>
      </tr>
      <tr>
        <th>RNAseq-uninfected-1</th>
        <td>
          <a href="sequences/RNAseq-uninfected-1.filtered.fa.gz">RNAseq-uninfected-1.fa.gz</a><br/>
          <a href="sequences/RNAseq-uninfected-1.filtered.fq.gz">RNAseq-uninfected-1.fq.gz</a>
        </td>
        <td>7.4 MiB<br/>3.7 MiB</td>
        <td><a href="http://genetics.bwh.harvard.edu/pph/FASTA.html">Fasta</a> (collapsed)<br/><a href="http://maq.sourceforge.net/fastq.shtml">FastQ</a></td>
      </tr>
    </tbody>
  </table>

  <h3><a name="data_alignment_files" href="#data_alignment_files">Alignment Files</a></h3>
  <table>
    <thead>
      <tr>
        <th>Sample</th>
        <th>File</th>
        <th>Size</th>
        <th>Format</th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <th>CLIP-35L33G</th>
        <td>
          <a href="alignments/CLIP-35L33G.sorted.bam">CLIP-35L33G.sorted.bam</a><br/>
          <a href="alignments/CLIP-35L33G.sorted.bam">CLIP-35L33G.sorted.bam.bai</a>
        </td>
        <td>7.4 MiB<br/>0.7 KiB</td>
        <td><a href="https://genome.ucsc.edu/FAQ/FAQformat.html#format5.1">bam</a> (sorted)<br/>bam index</td>
      </tr>
      <tr>
        <th>CLIP-2J3</th>
        <td>
          <a href="alignments/CLIP-2J3.sorted.bam">CLIP-2J3.sorted.bam</a><br/>
          <a href="alignments/CLIP-2J3.sorted.bam">CLIP-2J3.sorted.bam.bai</a>
        </td>
        <td>7.4 MiB<br/>0.7 KiB</td>
        <td>bam (sorted)<br/>bam index</td>
      </tr>
      <tr>
        <th>RNAseq-uninfected-1</th>
        <td>
          <a href="alignments/RNAseq-uninfected-1.sorted.bam">RNAseq-uninfected-1.sorted.bam</a><br/>
          <a href="alignments/RNAseq-uninfected-1.sorted.bam">RNAseq-uninfected-1.sorted.bam.bai</a>
        </td>
        <td>7.4 MiB<br/>0.7 KiB</td>
        <td>bam (sorted)<br/>bam index</td>
      </tr>
    </tbody>
  </table>

  <h3><a name="data_annotations" href="#data_annotations">Sequence Annotation Files</a></h3>
  <table>
    <thead>
      <tr>
        <th>Sample</th>
        <th>File</th>
        <th>Size</th>
        <th>Format</th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <th>CLIP-35L33G</th>
        <td>
          <a href="annotations/CLIP-35L33G.anno.gz">CLIP-35L33G.anno.gz</a>
        </td>
        <td>9.1 MiB</td>
        <td><a href="report/help/annotations.html" target="_new">(help)</a></td>
      </tr>
      <tr>
        <th>CLIP-2J3</th>
        <td>
          <a href="annotations/CLIP-2J3.anno.gz">CLIP-2J3.anno.gz</a>
        </td>
        <td>9.1 MiB</td>
        <td><a href="report/help/annotations.html" target="_new">(help)</a></td>
      </tr>
      <tr>
        <th>RNAseq-uninfected-1</th>
        <td>
          <a href="annotations/RNAseq-uninfected-1.anno.gz">RNAseq-uninfected-1.anno.gz</a>
        </td>
        <td>9.1 MiB</td>
        <td><a href="report/help/annotations.html" target="_new">(help)</a></td>
      </tr>
    </tbody>
  </table>

  <h3><a name="data_site_catalogs" href="#data_site_catalogs">
    Nucleotide-resolution Statistics for Binding Site Identification</a></h3>
  <table>
    <thead>
      <tr>
        <th>Sample</th>
        <th>File</th>
        <th>Size</th>
        <th>Format</th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <th>CLIP-35L33G</th>
        <td>
          <a href="erroranalyses/CLIP-35L33G.catalog.gz">CLIP-35L33G.catalog.gz</a>
        </td>
        <td>59.1 MiB</td>
        <td><a href="report/help/site_catalogs.html" target="_new">(help)</a></td>
      </tr>
      <tr>
        <th>CLIP-2J3</th>
        <td>
          <a href="catalogtations/CLIP-2J3.catalog.gz">CLIP-2J3.catalog.gz</a>
        </td>
        <td>9.1 MiB</td>
        <td><a href="report/help/site_catalogs.html" target="_new">(help)</a></td>
      </tr>
      <tr>
        <th>RNAseq-uninfected-1</th>
        <td>
          <a href="catalogtations/RNAseq-uninfected-1.catalog.gz">RNAseq-uninfected-1.catalog.gz</a>
        </td>
        <td>9.1 MiB</td>
        <td><a href="report/help/site_catalogs.html" target="_new">(help)</a></td>
      </tr>
    </tbody>
  </table>

  <h3><a name="data_site_fdr_estimation" href="#data_site_fdr_estimation">
    Binding Site Score Cutoff Tables for Desired FDR Levels</a></h3>
  <table>
    <thead>
      <tr>
        <th>Sample</th>
        <th>File</th>
        <th>Size</th>
        <th>Format</th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <th>CLIP-35L33G</th>
        <td>
          <a href="erroranalyses/CLIP-35L33G.del.fdr_cutoffs">CLIP-35L33G.del.fdr_cutoffs</a><br/>
          <a href="erroranalyses/CLIP-35L33G.mod.fdr_cutoffs">CLIP-35L33G.mod.fdr_cutoffs</a><br/>
          <a href="erroranalyses/CLIP-35L33G.moddel.fdr_cutoffs">CLIP-35L33G.moddel.fdr_cutoffs</a><br/>
          <a href="erroranalyses/CLIP-35L33G.entropy.fdr_cutoffs">CLIP-35L33G.entropy.fdr_cutoffs</a><br/>
          <a href="erroranalyses/CLIP-35L33G.t2c.fdr_cutoffs">CLIP-35L33G.t2c.fdr_cutoffs</a>
        </td>
        <td>59.1 KiB<br/>12.5 KiB<br/>12.5 KiB<br/>12.5 KiB<br/>12.5 KiB</td>
        <td><a href="report/help/site_catalogs.html" target="_new">(help)</a></td>
      </tr>
      <tr>
        <th>CLIP-2J3</th>
        <td>
          <a href="erroranalyses/CLIP-2J3.del.fdr_cutoffs">CLIP-2J3.del.fdr_cutoffs</a><br/>
          <a href="erroranalyses/CLIP-2J3.mod.fdr_cutoffs">CLIP-2J3.mod.fdr_cutoffs</a><br/>
          <a href="erroranalyses/CLIP-2J3.moddel.fdr_cutoffs">CLIP-2J3.moddel.fdr_cutoffs</a><br/>
          <a href="erroranalyses/CLIP-2J3.entropy.fdr_cutoffs">CLIP-2J3.entropy.fdr_cutoffs</a><br/>
          <a href="erroranalyses/CLIP-2J3.t2c.fdr_cutoffs">CLIP-2J3.t2c.fdr_cutoffs</a>
        </td>
        <td>59.1 KiB<br/>12.5 KiB<br/>12.5 KiB<br/>12.5 KiB<br/>12.5 KiB</td>
        <td><a href="report/help/site_catalogs.html" target="_new">(help)</a></td>
      </tr>
      <tr>
        <th>RNAseq-uninfected-1</th>
        <td>
          <a href="erroranalyses/RNAseq-uninfected-1.del.fdr_cutoffs">RNAseq-uninfected-1.del.fdr_cutoffs</a><br/>
          <a href="erroranalyses/RNAseq-uninfected-1.mod.fdr_cutoffs">RNAseq-uninfected-1.mod.fdr_cutoffs</a><br/>
          <a href="erroranalyses/RNAseq-uninfected-1.moddel.fdr_cutoffs">RNAseq-uninfected-1.moddel.fdr_cutoffs</a><br/>
          <a href="erroranalyses/RNAseq-uninfected-1.entropy.fdr_cutoffs">RNAseq-uninfected-1.entropy.fdr_cutoffs</a><br/>
          <a href="erroranalyses/RNAseq-uninfected-1.t2c.fdr_cutoffs">RNAseq-uninfected-1.t2c.fdr_cutoffs</a>
        </td>
        <td>59.1 KiB<br/>12.5 KiB<br/>12.5 KiB<br/>12.5 KiB<br/>12.5 KiB</td>
        <td><a href="report/help/site_catalogs.html" target="_new">(help)</a></td>
      </tr>
    </tbody>
  </table>

  <h3><a name="data_piranha_peaks" href="#data_piranha_peaks">
    Alternative Binding Sites Calls (Piranha)</a></h3>
  <table>
    <thead>
      <tr>
        <th>Sample</th>
        <th>File</th>
        <th>Size</th>
        <th>Format</th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <th>CLIP-35L33G</th>
        <td>
          <a href="peaks/CLIP-35L33G.piranha.gz">CLIP-35L33G.piranha.gz</a>
        </td>
        <td>59.1 MiB</td>
        <td><a href="https://genome.ucsc.edu/FAQ/FAQformat.html#format1">gzipped bed</a></td>
      </tr>
      <tr>
        <th>CLIP-2J3</th>
        <td>
          <a href="peaks/CLIP-2J3.piranha.gz">CLIP-2J3.piranha.gz</a>
        </td>
        <td>9.1 MiB</td>
        <td><a href="https://genome.ucsc.edu/FAQ/FAQformat.html#format1">gzipped bed</a></td>
      </tr>
      <tr>
        <th>RNAseq-uninfected-1</th>
        <td>
          <a href="peaks/RNAseq-uninfected-1.piranha.gz">RNAseq-uninfected-1.piranha.gz</a>
        </td>
        <td>9.1 MiB</td>
        <td><a href="https://genome.ucsc.edu/FAQ/FAQformat.html#format1">gzipped bed</a></td>
      </tr>
    </tbody>
  </table>

  <h3><a name="data_enriched_transcripts" href="#data_enriched_transcripts">
    Target Transcript List Based on Tag Abundance</a></h3>
  <table>
    <thead>
      <tr>
        <th>Pair Set</th>
        <th>File</th>
        <th>Size</th>
        <th>Format</th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <th>CLIP-35L33G</th>
        <td>
          <a href="stats/CLIP-35L33G.enrichment.csv">CLIP-35L33G.enrichment.csv</a>
        </td>
        <td>59.1 MiB</td>
        <td><a href="https://genome.ucsc.edu/FAQ/FAQformat.html#format1">csv</a></td>
      </tr>
      <tr>
        <th>CLIP-2J3</th>
        <td>
          <a href="stats/CLIP-2J3.enrichment.csv">CLIP-2J3.enrichment.csv</a>
        </td>
        <td>9.1 MiB</td>
        <td><a href="https://genome.ucsc.edu/FAQ/FAQformat.html#format1">csv</a></td>
      </tr>
    </tbody>
  </table>

  <h3><a name="data_genome_arrays" href="#data_genome_arrays">
    Genome-Spanning Arrays of CLIP Tag Counts (and more)</a></h3>
  <table>
    <thead>
      <tr>
        <th>Sample</th>
        <th>File</th>
        <th>Size</th>
        <th>Format</th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <th>CLIP-35L33G</th>
        <td>
          <a href="quantitations/CLIP-35L33G.gspace">CLIP-35L33G.gspace</a> (dir)
        </td>
        <td>59.1 MiB</td>
        <td><a href="https://genome.ucsc.edu/FAQ/FAQformat.html#format1">(help)</a></td>
      </tr>
      <tr>
        <th>CLIP-2J3</th>
        <td>
          <a href="quantitations/CLIP-2J3.gspace">CLIP-2J3.gspace</a> (dir)
        </td>
        <td>59.1 MiB</td>
        <td><a href="https://genome.ucsc.edu/FAQ/FAQformat.html#format1">(help)</a></td>
      </tr>
      <tr>
        <th>RNAseq-uninfected-1</th>
        <td>
          <a href="quantitations/RNAseq-uninfected-1.gspace">RNAseq-uninfected-1.gspace</a> (dir)
        </td>
        <td>59.1 MiB</td>
        <td><a href="https://genome.ucsc.edu/FAQ/FAQformat.html#format1">(help)</a></td>
      </tr>
    </tbody>
  </table>
</div>
{#-
vim: sts=2 sw=2 et ft=htmljinja
-#}
